File: flutter_inappwebview_android/android/src/main/java/com/pichillilorenzo/flutter_inappwebview_android/webview/WebViewInstanceRegistry.java
package com.pichillilorenzo.flutter_inappwebview_android.webview;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.pichillilorenzo.flutter_inappwebview_android.webview.in_app_webview.InAppWebView;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public final class WebViewInstanceRegistry {
  private static final Map<String, WeakReference<InAppWebView>> REGISTRY = new ConcurrentHashMap<>();

  private WebViewInstanceRegistry() {
  }

  public static void register(@NonNull String instanceId, @NonNull InAppWebView webView) {
    removeStaleReferencesFor(webView);
    REGISTRY.put(instanceId, new WeakReference<>(webView));
  }

  public static void unregister(@Nullable String instanceId, @Nullable InAppWebView webView) {
    if (instanceId == null) {
      return;
    }
    WeakReference<InAppWebView> reference = REGISTRY.get(instanceId);
    InAppWebView registered = reference != null ? reference.get() : null;
    if (registered == null || registered == webView) {
      REGISTRY.remove(instanceId);
    }
  }

  public static void unregister(@NonNull InAppWebView webView) {
    removeStaleReferencesFor(webView);
  }

  @Nullable
  public static InAppWebView get(@NonNull String instanceId) {
    WeakReference<InAppWebView> reference = REGISTRY.get(instanceId);
    InAppWebView webView = reference != null ? reference.get() : null;
    if (webView == null && reference != null) {
      REGISTRY.remove(instanceId);
    }
    return webView;
  }

  @NonNull
  public static List<String> getRegisteredInstanceIds() {
    cleanUp();
    return new ArrayList<>(REGISTRY.keySet());
  }

  public static boolean isRegistered(@NonNull String instanceId) {
    return get(instanceId) != null;
  }

  public static void clear() {
    REGISTRY.clear();
  }

  private static void cleanUp() {
    List<String> keysToRemove = new ArrayList<>();
    for (Map.Entry<String, WeakReference<InAppWebView>> entry : REGISTRY.entrySet()) {
      if (entry.getValue().get() == null) {
        keysToRemove.add(entry.getKey());
      }
    }
    for (String key : keysToRemove) {
      REGISTRY.remove(key);
    }
  }

  private static void removeStaleReferencesFor(@NonNull InAppWebView webView) {
    List<String> keysToRemove = new ArrayList<>();
    for (Map.Entry<String, WeakReference<InAppWebView>> entry : REGISTRY.entrySet()) {
      InAppWebView registered = entry.getValue().get();
      if (registered == null || registered == webView) {
        keysToRemove.add(entry.getKey());
      }
    }
    for (String key : keysToRemove) {
      REGISTRY.remove(key);
    }
  }
}

File: flutter_inappwebview_ios/ios/Classes/InAppWebView/WebViewInstanceRegistry.swift
import Foundation

final class WebViewInstanceRegistry {
    private struct WeakInAppWebView {
        weak var value: InAppWebView?
    }

    private static var registry: [String: WeakInAppWebView] = [:]
    private static let lock = NSObject()

    private init() {}

    static func register(instanceId: String, webView: InAppWebView) {
        objc_sync_enter(lock)
        removeStaleReferencesLocked(for: webView)
        registry[instanceId] = WeakInAppWebView(value: webView)
        objc_sync_exit(lock)
    }

    static func unregister(instanceId: String, webView: InAppWebView?) {
        objc_sync_enter(lock)
        if let entry = registry[instanceId] {
            if let registered = entry.value {
                if webView == nil || registered === webView {
                    registry.removeValue(forKey: instanceId)
                }
            } else {
                registry.removeValue(forKey: instanceId)
            }
        }
        objc_sync_exit(lock)
    }

    static func unregister(webView: InAppWebView) {
        objc_sync_enter(lock)
        removeStaleReferencesLocked(for: webView)
        objc_sync_exit(lock)
    }

    static func get(instanceId: String) -> InAppWebView? {
        objc_sync_enter(lock)
        let webView = registry[instanceId]?.value
        if webView == nil {
            registry.removeValue(forKey: instanceId)
        }
        objc_sync_exit(lock)
        return webView
    }

    static func getRegisteredInstanceIds() -> [String] {
        objc_sync_enter(lock)
        cleanUpLocked()
        let keys = Array(registry.keys)
        objc_sync_exit(lock)
        return keys
    }

    static func isRegistered(instanceId: String) -> Bool {
        return get(instanceId: instanceId) != nil
    }

    static func clear() {
        objc_sync_enter(lock)
        registry.removeAll()
        objc_sync_exit(lock)
    }

    private static func cleanUpLocked() {
        let keysToRemove = registry.compactMap { (key: String, value: WeakInAppWebView) -> String? in
            return value.value == nil ? key : nil
        }
        for key in keysToRemove {
            registry.removeValue(forKey: key)
        }
    }

    private static func removeStaleReferencesLocked(for webView: InAppWebView) {
        let keysToRemove = registry.compactMap { (key: String, value: WeakInAppWebView) -> String? in
            guard let registered = value.value else {
                return key
            }
            return registered === webView ? key : nil
        }
        for key in keysToRemove {
            registry.removeValue(forKey: key)
        }
    }
}